Module system deep dive tutorial (#645)
NixOS/nix.dev: source/tutorials/module-system/module-system.md

Reviewer: Something I've felt as I'm going through this is that I have no sense of where I am in relation to the overall goals of the project i.e. as I'm defining a particular option, how much is there left to do in this tutorial. It feels like a never ending list of "now we're going to do this". There should be some kind of overview at the beginning so that the user can put each step into a larger context.

Reviewer 2: Added introductory notes to each sections.

# Module system introduction

> 

Reviewer 2: Actually this is really good, just reorder and reword slightly.

```suggestion
Much of the power in NixOS comes from the module system.
It provides mechanisms for conveniently declaring and automatically merging interdependent attribute sets, making it easy to compose configurations in a type-safe way.

In this tutorial you'll learn what a module is and how to define one, what options are and how to declare them, how to express dependencies between modules, and a practical way to use Nix to wrap external APIs.

You'll write your first modules to interact with the Google Maps API, declaring options which represent map geometry, location pins, and more.

Be prepared to see some Nix errors: during the tutorial, you will first write some *incorrect* configurations, creating opportunities to discuss the resulting error messages and how to resolve them, particularly when discussing type checking.
```

> Note: This tutorial was created from https://github.com/tweag/summer-of-nix-modules, as presented by @infinisil at Summer of Nix 2021, presentation can be seen here (might not work in browser, `mpv` should work): https://infinisil.com/modules.mp4

Reviewer 3: https://github.com/NixOS/nix.dev/issues/785

## Empty module

The simplest module you can have is just an empty attribute set, which
as you might expect, doesn't do anything!

```diff
diff --git a/default.nix b/default.nix
new file mode 100644
index 0000000..1797133
--- /dev/null
+++ b/default.nix
@@ -0,0 +1,3 @@
+{
+  
+}
```

## Module arguments: lib

Modules can be just an attribute set. But if you want access to some
arguments you need to change it into a function taking an attribute set
with an ellipsis (that's the "..."). In this case we only match on the
`lib` argument, which gives us access to nixpkgs library functions.

Note that the ellipsis is necessary since arbitrary arguments can be
passed to modules.

```diff
diff --git a/default.nix b/default.nix
index 1797133..f7569bd 100644
--- a/default.nix
+++ b/default.nix
@@ -1,3 +1,3 @@
-{
+{ lib, ... }: {
   
 }

> ```

Reviewer 2: Not sure if we should say "matching", as the more functionally minded people may read this as pattern matching. This is not even remotely the case.

In general, I was just recently thinking about it: While it's true that there are multiple "syntaxes" for the module system, we probably should teach their existence as an advanced step. How about we stick to `{ ... }: { }` as the main form? I don't think it's worthwhile showing an empty attrset just [for the sake of simplicity](https://m.xkcd.com/2587/).

## Declaring generate.script option

In order for modules to be useful, we need to have options, so let's
declare one. Options are declared by defining an attribute with
`lib.mkOption` under the `options` attribute. Here we're defining the
option `generate.script`.

> While there are many attributes to customize options, the most

Reviewer 4: Where I can find those other attributes?

Author: Afaik the source has the only complete listing: https://github.com/NixOS/nixpkgs/blob/fa1d4611cc2ce41d7b93c4a8e8db0efaaf286475/lib/options.nix#L66-L88

important one is `type`, which specifies what values are valid for an
option, and how/whether multiple values should be merged together.

In the nixpkgs library there are a number of types available under
`lib.types`. Here we're using the `lines` type, which specifies that:
- Only strings are valid values
- Multiple strings are joined with newlines

```diff
diff --git a/default.nix b/default.nix
index f7569bd..cb423a9 100644
--- a/default.nix
+++ b/default.nix
@@ -1,3 +1,9 @@
 { lib, ... }: {
+
+  options = {
+    generate.script = lib.mkOption {
+      type = lib.types.lines;
+    };
+  };
   
 }
```

Let's try to evaluate this with this file:

```nix
# eval.nix
(import <nixpkgs/lib>).evalModules {
  modules = [
    ./default.nix
  ];
}
```

Then we run
```bash
nix-instantiate --eval eval.nix -A config.generate.script
```

Trying to evaluate the `generate.script` option however, we get an error

> that the option is used but not defined, indicating that we need to

Reviewer 4: This always confounded me:
- `= config.` _used_: basic¹ user may never see this
- `config. =` _defined_: to his POV this is where he _used_ or _defined_ it
- `options. =` _defined?_: to me, this is where I defined it (I'm not basic user)

_¹basic_:
lower basic nix: never saw a nix file before :baby: 
basic nix modules use: `config. = ""`, use existing options with static values
mid nix modules use: `config. = config.`, use existing options reusing existing values
upper mid nix modules use: `config. = lib.mkIf config.`, use existing options using some kind of logic
higher nix modules use: `options. =`, define his own options

Author: In the module system, `config = ` are usually called definitions, while `options = ` are called declarations

actually give a value to the option.

## Type checking: Assigning integer to generate.script

We can try to assign an integer to our option of type `lines`, but the
module system correctly throws an error saying that our definition
doesn't match the options type.

```diff
diff --git a/default.nix b/default.nix
index cb423a9..0a9162e 100644
--- a/default.nix
+++ b/default.nix
@@ -5,5 +5,9 @@
       type = lib.types.lines;
     };
   };
+
+  config = {
+    generate.script = 42;
+  };
   
 }
```

## Successful evaluation: Assigning a string to generate.script

We can make type checking pass by assigning a string to the option,
giving us our first successful evaluation of the `generate.script`
option.

In this case, we assign a script which calls the Google Maps Static API

> to generate a world map, then displaying the result using icat

Reviewer: ```suggestion
to generate a world map, then displaying the result using `icat`
```

(image-cat), both of which are helper scripts.

```diff
diff --git a/default.nix b/default.nix
index 0a9162e..24f9c34 100644
--- a/default.nix
+++ b/default.nix
@@ -7,7 +7,9 @@
   };
 
   config = {
-    generate.script = 42;
+    generate.script = ''
+      map size=640x640 scale=2 | icat
+    '';
   };
   
 }
```

TODO: Create derivations to get these commands

## A new list option: Declaring generate.requestParams

> Let's introduce another option, generate.requestParams. For its type,

Reviewer: ```suggestion
Let's introduce another option, `generate.requestParams`. For its type,
```

we'll use `listOf <nestedType>`, which is a generic list type where each
element has to match the given nested type. In our case we want `str` to
be the nested type, which is a generic string type.

Note that the difference between `str` and `lines` is in their merging
behavior:
- For `lines`, multiple definitions get merged by concatenation with
  newlines
- For `str`, multiple definitions are not allowed. Which in this case is
  mostly irrelevant however, since we can't really define a list element

>   multiple times.

Reviewer 4: Maybe we should be using _str_ instead of _lines_ at line 70, usually, scripts aren't concatenated anyway.
Then we could move this to line 56
It would reduce complexity (number of concepts) in section of lists.

Author: I intentionally tried to introduce as many different types as possible to show what's there. I think it's somewhat common for scripts to be concatenated though, e.g. `systemd.services.<name>.{script,preStart}` is like that

Reviewer 4: > I intentionally tried to introduce as many different types as possible to show what's there.

Agree with this, but we are in a list section, reader have to bash his head against lists, e we're talking about strings,
Ok that computationally speaking a string is a list of chars, and lines is internally represented as a list of strings. But no one needs to know it until the next level. ;-)

Other problem of this, I usually read documentations again trying to find concepts that I know they exist in that document like a reference/dictionary. Include string in list, would be like including the definition of some S word in L page.

That is why suggested move this to line 56 (where we introduce lines)
Also, I was thinking in add other string types strMatching and enum (ok is not a string but mostly used with string)

This suggestion also reduce our str definition that has this _Which in this case is mostly irrelevant however, since we can't really define a list element multiple times._ that may cause confusion, since we are telling something important and saying that it doesn't matter.

Reviewer: I think how it's written makes sense. Don't block on this. If anything I would just remove the entire "Which in this case ..." sentence since by it's own admission it's not terribly relevant.

```diff
diff --git a/default.nix b/default.nix
index 24f9c34..fd5027a 100644
--- a/default.nix
+++ b/default.nix
@@ -4,12 +4,21 @@
     generate.script = lib.mkOption {
       type = lib.types.lines;
     };
+
+    generate.requestParams = lib.mkOption {
+      type = lib.types.listOf lib.types.str;
+    };
   };
 
   config = {
     generate.script = ''
       map size=640x640 scale=2 | icat
     '';
+
+    generate.requestParams = [
+      "size=640x640"
+      "scale=2"
+    ];
   };
   
 }
```

## Dependencies between options: Using generate.requestParams

> A collection of modules generally only has a single option that is meant

Reviewer: This could be clearer. What I think you trying to say is that a user may only specify a single option, which causes several other options to then be evaluated (or something like that). When I read "a collection of modules generally only has a single option that is meant to be evaluated" I wonder how that can be the case. How can several modules all work together to provide a _single_ option? What's the point of the modules that don't contain that option? These are all questions a new user will have after reading this.

to be evaluated. That's the option that generates the final result we're
interested in, which in our case is `generate.script`.

In order to build up abstractions on that, we have the ability for
options to depend on other options. In this case we want the
`generate.script` option to use the values of `generate.requestParams`.
We can access the values of options by adding the `config` argument to
the argument list at the top and using e.g.

> `config.generate.requestParams` to access that options value.

Reviewer: You need to explain why this works the way that it does. This is the thought process of a new user in this scenario:
- I've defined `generate.script` and `generate.requestParams` next to each other in the `config` attribute.
- If they're next to each other why can't I refer to them? Can't I use `rec`?
- The correct way is to add a function argument named `config`, but I'm currently defining `config`. How on earth does that make sense?

We're then using `lib.concatStringsSep " "` to join each list element
of the option together into an argument list.

```diff
diff --git a/default.nix b/default.nix
index fd5027a..050a98e 100644
--- a/default.nix
+++ b/default.nix
@@ -1,4 +1,4 @@
-{ lib, ... }: {
+{ lib, config, ... }: {
 
   options = {
     generate.script = lib.mkOption {
@@ -12,7 +12,9 @@
 
   config = {
     generate.script = ''
-      map size=640x640 scale=2 | icat
+      map ${lib.concatStringsSep " "
+            config.generate.requestParams
+           } | icat
     '';
 
     generate.requestParams = [
```

## Conditional definitions: Introducing map.zoom option

We now introduce the new option `map.zoom` in order to control the zoom
level of the map. We'll use a new type for it, `nullOr <type>`, which
accepts the value `null`, but also values of its argument type. We're
using `null` here to mean an inferred zoom level.

For this option, we'll set a default value which should be used if it's

> not defined otherwise, which we can do using `mkOption`'s `default`

Reviewer: This isn't specific to this tutorial, but we need some kind of terminology for the attributes that make up an attribute set passed as an argument. For example, it's incorrect to say that `buildInputs` is an argument to `mkDerivation` since `mkDerivation` is a function that takes a single argument and that argument is an attrset. Maybe something like "input attribute"?

Reviewer 2: I suggest splitting that out into an issue. (GitHub apparently has a UI for that in the `...` menu on the comment.)

argument.

Now we want to use this option to define another element in
`generate.requestParams`, but we only want to add this element if its
value is non-null. We can do this using the `mkIf <condition>
<definition>` function, which only adds a definition if the condition
holds.

```diff
diff --git a/default.nix b/default.nix
index 050a98e..6b6e1e1 100644
--- a/default.nix
+++ b/default.nix
@@ -8,6 +8,13 @@
     generate.requestParams = lib.mkOption {
       type = lib.types.listOf lib.types.str;
     };
+
+    map = {
+      zoom = lib.mkOption {

> +        type = lib.types.nullOr lib.types.int;

Reviewer: How can this ever be null if there's a default value?

Reviewer 2: If it's explicitly set with `config.zoom = null;`

+        default = 2;
+      };
+    };
   };
 
   config = {
@@ -20,6 +27,8 @@
     generate.requestParams = [
       "size=640x640"
       "scale=2"
+      (lib.mkIf (config.map.zoom != null)
+        "zoom=${toString config.map.zoom}")
     ];
   };
   
```

## Declaring map.center option

> Similarly, let's declare a map.center option, declaring where the map

Reviewer: ```suggestion
Similarly, let's declare a `map.center` option, declaring where the map
```

should be centered.

> We'll be using a small utility for geocoding location names, aka turning

Reviewer: ```suggestion
We'll be using a small utility for geocoding location names, i.e. turning
```

       (lib.mkIf (config.map.zoom != null)
         "zoom=${toString config.map.zoom}")
+      (lib.mkIf (config.map.center != null)
+        "center=\"$(geocode ${
+          lib.escapeShellArg config.map.center
+        })\"")
     ];
   };
   
```

## Splitting modules: importing marker.nix

The module system allows you to split your config into multiple files
via the `imports` attribute, which can define further modules to import.
This allows us to logically separate options and config for different
parts.

Here we create a new module `marker.nix`, where we will declare options
for defining markers on the map

```diff
diff --git a/default.nix b/default.nix
index 098d135..9d9f29d 100644
--- a/default.nix
+++ b/default.nix
@@ -1,5 +1,9 @@
 { lib, config, ... }: {
 
+  imports = [
+    ./marker.nix
+  ];
+
   options = {
     generate.script = lib.mkOption {
       type = lib.types.lines;
diff --git a/marker.nix b/marker.nix
new file mode 100644
index 0000000..035c28d
--- /dev/null
+++ b/marker.nix
@@ -0,0 +1,3 @@
+{ lib, config, ... }: {
+
+}
```

## Submodule types: Declaring map.markers option

One of the most useful types of the module system is the `submodule`

> type. This type allows you to define a nested module system evaluation,

Reviewer: I don't know what these two sentences mean.

with its own options. Every value of such a type is then interpreted
(by default) as a `config` assignment of the nested module evaluation.

In this case we're defining a `map.markers` option, whose type is a list
of submodules with a nested `location` type, allowing us to define a
list of markers on the map, where each assignment is type checked
according to the submodule.

```diff
diff --git a/marker.nix b/marker.nix
index 035c28d..6a6c686 100644
--- a/marker.nix
+++ b/marker.nix
@@ -1,3 +1,20 @@
-{ lib, config, ... }: {
+{ lib, config, ... }:
+let
+
+  markerType = lib.types.submodule {
+    options = {
+      location = lib.mkOption {
+        type = lib.types.nullOr lib.types.str;
+        default = null;
+      };
+    };
+  };
+in {
+
+  options = {
+    map.markers = lib.mkOption {
+      type = lib.types.listOf markerType;
+    };
+  };
 
 }
```

> ## Single option namespace: Defining markers in generate.requestParams

Reviewer: I'm not sure what "single option namespace" means here.

> Since all modules in `imports` are treated the same, we can also freely

Reviewer: I'm not sure what this sentence means.

assign an option defined in our initial module. In this case we want to
add some request parameters derived from the `map.markers` option, so
that markers actually show up on the map.

```diff
diff --git a/marker.nix b/marker.nix
index 6a6c686..c2c5da2 100644
--- a/marker.nix
+++ b/marker.nix
@@ -17,4 +17,26 @@ in {
     };
   };
 
+  config = {
+
+    map.markers = [
+      { location = "new york"; }
+    ];
+
+    generate.requestParams = let
+      paramForMarker = marker:
+        let
+          attributes =
+            [
+              "$(geocode ${
+                lib.escapeShellArg marker.location
+              })"
+            ];

> +        in "markers=${

Reviewer: It's kind of hard to understand what's happening here on a language level. Are you joining each attribute with the `\\|` string?

+      (lib.length config.map.markers >= 1)
+      null;
+
+    map.zoom = lib.mkIf
+      (lib.length config.map.markers >= 2)
+      null;
+
     generate.requestParams = let
       paramForMarker = marker:
         let
```

## Nested submodules: Introducing users option

We will now introduce the option `users`, where we make use of a very
useful type, `lib.types.attrsOf <subtype>`. This type lets us specify an
attribute set as a value, where the keys can be arbitrary, but each
value has to conform to the given <subtype>.

In this case, we'll use another submodule as the subtype, one that
allows declaring a departure marker, which notably also makes use of our
`markerType` submodule, giving us a nested structure of submodules.

We're now propagating each of the users marker definitions to the
`map.markers` option.

```diff
diff --git a/marker.nix b/marker.nix
index b80ddd0..3c54ad8 100644
--- a/marker.nix
+++ b/marker.nix
@@ -9,9 +9,24 @@ let
       };
     };
   };
+
+  userType = lib.types.submodule {
+    options = {
+      departure = lib.mkOption {
+        type = markerType;
+        default = {};
+      };
+    };
+  };
+
 in {
 
   options = {
+

> +    users = lib.mkOption {

Reviewer: This is as far as I got before I mentally ran out of gas.

Reviewer 2: All of Nix is running on fumes, having gas is clearly a luxury.

use the same `generate.requestParams` to influence the API call to
include our defined paths

```diff
diff --git a/marker.nix b/marker.nix
index 2c0c1a8..ffb8185 100644
--- a/marker.nix
+++ b/marker.nix
@@ -56,6 +56,10 @@ let
 
 in {
 
+  imports = [
+    ./path.nix
+  ];
+
   options = {
 
     users = lib.mkOption {
diff --git a/path.nix b/path.nix
new file mode 100644
index 0000000..554a88b
--- /dev/null
+++ b/path.nix
@@ -0,0 +1,32 @@
+{ lib, config, ... }:
+let
+  pathType = lib.types.submodule {
+
+    options = {
+      locations = lib.mkOption {
+        type = lib.types.listOf lib.types.str;
+      };
+    };
+
+  };
+in {
+  options = {
+    map.paths = lib.mkOption {
+      type = lib.types.listOf pathType;
+    };
+  };
+
+  config = {
+    generate.requestParams = let
+      attrForLocation = loc:
+        "$(geocode ${lib.escapeShellArg loc})";
+      paramForPath = path:
+        let
+          attributes =
+            map attrForLocation path.locations;
+        in "path=${
+            lib.concatStringsSep "\\|" attributes
+          }";
+      in map paramForPath config.map.paths;
+  };
+}
```

> ## Arrival marker

Reviewer 5: ```suggestion
## The Arrival Marker
```

Now in order for users to be able to draw a path in their definitions,
we need to allow them to specify another marker. We will copy the
`departure` option declaration to a new `arrival` option for that.

```diff
diff --git a/marker.nix b/marker.nix
index ffb8185..940b8f8 100644
--- a/marker.nix
+++ b/marker.nix
@@ -46,11 +46,18 @@ let
         type = markerType;
         default = {};
       };
+
+      arrival = lib.mkOption {
+        type = markerType;
+        default = {};
+      };
     };
 
     config = {
       departure.style.label = lib.mkDefault
         (firstUpperAlnum name);
+      arrival.style.label = lib.mkDefault
+        (firstUpperAlnum name);
     };
   });
 
@@ -76,7 +83,7 @@ in {
     map.markers = lib.filter
       (marker: marker.location != null)
       (lib.concatMap (user: [
-        user.departure
+        user.departure user.arrival
       ]) (lib.attrValues config.users));
 
     map.center = lib.mkIf
```

> ## Connecting user paths

Reviewer 5: ```suggestion
## Connecting Markers by Paths
```

> # Deep dive demo: Wrapping the world in modules

Reviewer 3: This is too long to fit nicely in the sidebar (see deployed preview)

```suggestion
# Module system
```

Reviewer 2: I wouldn't want this to pretend to be the definitive module system tutorial. The title should also say what something about what to expect. (Also we can change names later on)

Reviewer 3: I mean, it's the *only* module system tutorial? I think I don't understand the argument.

Reviewer 2: ```suggestion
# Module system deep dive

Or: *Wrapping the world in modules*
```

Much of the power in Nixpkgs and NixOS comes from the module system.
It provides mechanisms for conveniently declaring and automatically merging interdependent attribute sets that follow dynamic type constraints, making it easy to express modular configurations.

## Overview

> This tutorial follows [@infinisil](https://github.com/infinisil)'s [presentation on modules](https://infinisil.com/modules.mp4)  ([source](https://github.com/tweag/summer-of-nix-modules)) for  participants of [Summer of Nix](https://github.com/ngi-nix/summer-of-nix) 2021.

Author: Fyi I just updated https://infinisil.com/modules.mp4 to use H.264 instead, so this should now easily play in browsers :)

It may help playing it back alongside the tutorial to better keep track of changes to the code you will work on.

### What will you learn?

In this tutorial you'll learn
- what a module is
- how to define one
- what options are
- how to declare them

> - how to express dependencies between modules

Reviewer 2: Meh, I'm not a fan of punctuation where visual structure already does the job.

During the tutorial, you will first write some *incorrect* configurations, creating opportunities to discuss the resulting error messages and how to resolve them, particularly when discussing type checking.

### What do you need?

- Familiarity with data types and general programming
- A {ref}`Nix installation <install-nix>` to run the examples
- Intermediate proficiency in reading and writing the Nix language

You will use two helper scripts for this exercise.
Download {download}`map <files/map>` and {download}`geocode <files/geocode>` to your working directory.

:::{warning}
To run the examples in this tutorial, you will need a [Google API key](https://developers.google.com/maps/documentation/maps-static/start#before-you-begin) in `$XDG_DATA_HOME/google-api/key`.
:::

### How long will it take?

This is a very long tutorial.
Prepare for at least 3 hours of work.

## Empty module

We have to start somewhere.
The simplest module is just a function that takes any attributes and returns an empty attribute set.

Write the following into a file called `default.nix`:

```nix

> # default.nix

Reviewer 2: Ah yeah thanks, I was hopelessly searching for the documentation because I dimly remembered that was a thing.

> # default.nix

Reviewer 3: https://github.com/NixOS/nix.dev/pull/786

This is done by declaring *options* that specify which values can be set and used elsewhere.
Options are declared by adding an attribute under the top-level `options` attribute, using `lib.mkOption`.

In this section, you will define the `scripts.output` option.

Change `default.nix` to include the following declaration:

```diff
# default.nix
 { lib, ... }: {

+ options = {
+   scripts.output = lib.mkOption {
+     type = lib.types.lines;
+   };
+ };

 }
```

While many attributes for customizing options are available, the most important one is `type`, which specifies which values are valid for an option.
There are several types available under [`lib.types`](https://nixos.org/manual/nixos/stable/#sec-option-types-basic) in the Nixpkgs library.

You have just declared `scripts.output` with the `lines` type, which specifies that the only valid values are strings, and that multiple definitions should be joined with newlines.

:::{note}
The name and attribute path of the option is arbitrary.
Here we use `scripts`, because we will add another script later, and call this one `output`, because it will output the resulting map.
:::

> ## Evaluating modules

Author: This section should be at the very start. `{ }` is not a module, it's just an attribute set. Only calling `evalModules` on it makes it be interpreted as module. So it's weird to talk about modules before `evalModules` is introduced.

Add the `center` option now, possibly with your own location as default value:

```diff
# default.nix
         type = lib.types.nullOr lib.types.int;
         default = 2;
       };
+
+      center = lib.mkOption {
+        type = lib.types.nullOr lib.types.str;
+        default = "switzerland";
+      };
     };
   };
```

To implement this behavior, you will use the {download}`geocode <files/geocode>` utility, which turns location names into coordinates.
There are multiple ways of making a new package accessible, but as an exercise, you will add it as an option in the module system.

First, add a new option to accommodate the package:

```diff
# default.nix
   options = {
     scripts.output = lib.mkOption {
       type = lib.types.package;
     };
+
+    scripts.geocode = lib.mkOption {
+      type = lib.types.package;
+    };
```

Then define the value for that option where you make the raw script reproducible by wrapping a call to it in `writeShellApplication`:

```diff
# default.nix
   config = {
+    scripts.geocode = pkgs.writeShellApplication {
+      name = "geocode";
+      runtimeInputs = with pkgs; [ curl jq ];

> +      text = "exec ${./geocode}";

Author: Might as well fetch this (and `map`) using Nix, so there's no need for a manual download :)

In the `let` block:

```diff
# marker.nix
         type = lib.types.nullOr lib.types.str;
         default = null;
       };
+
+      style.label = lib.mkOption {
+        type = lib.types.nullOr
+          (lib.types.strMatching "[A-Z0-9]");
+        default = null;
+      };
     };
   };
```

Again, `types.nullOr` allows for `null` values, and the default has been set to `null`.

In the `paramForMarker` function:

```diff
# marker.nix
       paramForMarker = marker:
         let
           attributes =
-            [
+            lib.optional
+              (marker.style.label != null)
+              "label:${marker.style.label}"
+            ++ [
               "$(geocode ${
                 lib.escapeShellArg marker.location
               })"
```

Here, the label for each `marker` is only propagated to the CLI parameters if `marker.style.label` is set.

## Defining a Default Label

Right now, if a label is not explicitly set, none will show up.
But since every `users` attribute has a name, we could use that as an automatic value instead.

> This `firstUpperAlnum` function allows you to retrieve the first character of the username, with the correct type for passing to `departure.style.label`:

Reviewer 3: What does `Alnum` mean?

     (lib.toUpper str));

+  # Either a color name or `0xRRGGBB`
+  colorType = lib.types.either
+    (lib.types.strMatching "0x[0-9A-F]{6}")
+    (lib.types.enum [
+      "black" "brown" "green" "purple" "yellow"
+      "blue" "gray" "orange" "red" "white" ]);
+
   markerType = lib.types.submodule {
     options = {
       location = lib.mkOption {
```

This allows either strings that matche a 24-bit hexadecimal number or are equal to one of the specified color names.

At the bottom of the `let` block, add the `style.color` option and specify a default value:

```diff
# marker.nix
           (lib.types.strMatching "[A-Z0-9]");
         default = null;
       };
+
+      style.color = lib.mkOption {
+        type = colorType;
+        default = "red";
+      };
     };
   };
```

Now add an entry to the `paramForMarker` list which makes use of the new option:

```diff
# marker.nix
               (marker.style.label != null)
               "label:${marker.style.label}"
             ++ [
+              "color:${marker.style.color}"
               "$(geocode ${
                 lib.escapeShellArg marker.location
               })"
```

## Marker Styling: Size

> In case you set many different markers, it would be helpful to have the ability to change their size individually.

Reviewer 3: Because this introduces `enum` by itself, I would move this up, before Color, which introduces `enum` inside an `either`

       paramForMarker = marker:
         let
+          size = {
+            tiny = "tiny";
+            small = "small";
+            medium = "mid";
+            large = null;
+          }.${marker.style.size};
+
```

Finally, add another `lib.optional` call to the `attributes` string, making use of the selected size:

```
# marker.nix
           attributes =
             lib.optional
               (marker.style.label != null)
               "label:${marker.style.label}"
+            ++ lib.optional
+              (size != null)
+              "size:${size}"
             ++ [
               "color:${marker.style.color}"
               "$(geocode ${
```

## The `pathType` Submodule

So far, you've created an option for declaring a *destination* marker, as well as several options for configuring the marker's visual representation.

Now we want to compute and display a route from the user's location to some destination.

The new option defined in the next section will allow you to set an *arrival* marker, which together with a destination allows you to draw *paths* on the map using the new module defined below.

To start, create a new `path.nix` file with the following contents:

> ```diff

Reviewer 3: ```suggestion
```nix
```

# path.nix
{ lib, config, ... }:
let
  pathType = lib.types.submodule {
    options = {
      locations = lib.mkOption {
        type = lib.types.listOf lib.types.str;
      };
    };
  };
in {
  options = {
    map.paths = lib.mkOption {
      type = lib.types.listOf pathType;
    };
  };

  config = {
    requestParams = let
      attrForLocation = loc:
        "$(geocode ${lib.escapeShellArg loc})";
      paramForPath = path:
        let
          attributes =
            builtins.map attrForLocation path.locations;
        in "path=${
            lib.concatStringsSep "\\|" attributes
          }";
      in builtins.map paramForPath config.map.paths;
  };
}
```

> The `path.nix` module defines an option for declaring a list of paths on our `map`, where each path is a list of strings for geographic locations.

Reviewer 3: This is where the defining/declaring terminology gets confusing.

```suggestion
The `path.nix` module declares an option for defining a list of paths on our `map`, where each path is a list of strings for geographic locations.
```

+        user.arrival.location
+      ];
+    }) (lib.filter (user:
+      user.departure.location != null
+      && user.arrival.location != null
+    ) (lib.attrValues config.users));
+
     requestParams = let
       attrForLocation = loc:
         "$(geocode ${lib.escapeShellArg loc})";
```

The new `map.paths` attribute contains a list of all valid paths defined for all users.

A path is valid only if the `departure` and `arrival` attributes are set for that user.

## Path Styling: Weight

Your users have spoken, and they demand the ability to customize the styles of their paths with a `weight` option.

As before, you'll now declare a new submodule for the path style.

While you could also directly declare the `style.weight` option, in this case you should use the submodule to be able reuse the path style type later.

Add the `pathStyleType` submodule option to the `let` block in `path.nix`:
```diff
# path.nix
 { lib, config, ... }:
 let
+
+  pathStyleType = lib.types.submodule {
+    options = {
+      weight = lib.mkOption {
+        type = lib.types.ints.between 1 20;
+        default = 5;
+      };
+    };
+  };
+
   pathType = lib.types.submodule {
```

:::{note}

> The `ints.between <lower> <upper>` type allows integers in the given (inclusive) range.

Reviewer 3: This should be part of the normal flow of the text, similar to every other type we introduced.
